### ISSUE TITLE
awaiting an arrow function needs extra parenthesis

### ISSUE THREAD

**Issue by omidkrad** [Microsoft/TypeScript#4885]
Minor issue. I think since we support

```typescript
function a() {
    return r => {}; // no need to put 'r => {}' in parentheses
}
```

we should also be able to do

```typescript
async w() {
    await resolve => setTimeout(resolve, 0);
}
```

the following works though

```typescript
async w() {
    await (resolve => setTimeout(resolve, 0));
}
```

P.S: async/await is awesome! I would be pissed if it does not land in ES7.

**Comment by SaschaNaz** [comment]
MS Edge with experimental flag doesn't complain with latter.

**Comment by mhegazy** [comment]
//cc: @rbuckton

**Comment by omidkrad** [comment]
#4904 may be related.

**Comment by rbuckton** [comment]
Using the `async` keyword with a simple arrow function was only recently added to the spec proposal. We will very likely be adding this soon.

**Comment by danquirk** [comment]
Calling this a bug since it just sounds like a task but suggestion/in discussion would be fine.

**Comment by omidkrad** [comment]
Out of curiosity, what would be the difference between

```typescript
async w() {
    await resolve => setTimeout(resolve, 0);
}
```

and

```typescript
async w() {
    return resolve => setTimeout(resolve, 0);
}
```

?

**Comment by omidkrad** [comment]
I guess in the first one we're not returning anything. So maybe the correct code is `return await`?

```typescript
async w() {
    return await resolve => setTimeout(resolve, 0);
}
```

That would be easy to do wrong.

**Comment by rbuckton** [comment]
The difference is that `await` is not a reserved word in ES6 or earlier.

**Comment by rbuckton** [comment]
@omidkrad What is the example proposing to do? Is your async `w()` function expecting to return a function?

Since `resolve => setTimeout(resolve, 0)` is not a Promise, the following two statements are roughly identical:

```ts
return await (resolve) => setTimeout(resolve, 0);
return resolve => setTimeout(resolve, 0);
```

If you're intent was to have `w()` return a promise that is resolved later, you would instead write:
```ts
function w() { // no `async` 
  return new Promise(resolve => setTimeout(resolve, 0);
}
async function z() {
  await w();
}
```

**Comment by omidkrad** [comment]
I think `await` should run a promise and return the result. So I think awaiting something that is not a Promise should produce error. For example: `var a = await 12` doesn't make much sense. I think awaiting a plain function is the same, unless the function is treated as a Promise. That is:

```typescript
async function w() {
    await resolve => setTimeout(resolve, 0);
}
```

IMO should be the same as

```typescript
async function w() {
    return new Promise(resolve => setTimeout(resolve, 0));
}
```

this will allow to completely abstract use of Promise through async/await. If you want to actually return a function, you would use the `return` keyword instead.

**Comment by weswigham** [comment]
@omidkrad We plan to follow the `async/await` ECMAScript specification. If you'd like to change the behavior of the async and await keywords in some way, perhaps you should get involved over at [the spec repo](https://github.com/tc39/ecmascript-asyncawait).

**Comment by omidkrad** [comment]
OK, this issue is particularly about parenthesis.

**Comment by omidkrad** [comment]
Submission to tc39 on the side discussion: https://github.com/tc39/ecmascript-asyncawait/issues/64

**Comment by rbuckton** [comment]
Per the [proposal](http://tc39.github.io/ecmascript-asyncawait/#prod-AwaitExpression):

*AwaitExpression*<sub>\[Yield]</sub>&emsp;**:**&emsp;`await`&emsp;*UnaryExpression*<sub>\[?Yield, Await]</sub>

The *ArrowFunction* production is part of *AssignmentExpression*, which is a much lower precedence than *UnaryExpression*. If you want to `await` an arrow function you have to wrap it in parentheses.

> `await` waits for a Promise and returns the result. So awaiting something that is not a Promise should produce error. For example: `var a = await 12` doesn't make much sense. I think awaiting a plain function or arrow function is the same, unless the function is treated as a Promise. 

While `var a = await 12` may not make much sense, what about `var a = await x`? Especially if `x` has the type `number | Promise<number>`? The proposal does not differentiate between a promise or a non-promise value. If you `await` a non-promise value, then your execution will resume with the same value in the next turn. The same goes for `return` in an async function. You can return a non-promise value or a promise value. If you return a promise value, its state will be adopted by the promise created by the async function.

As a result, the following code samples (ES.next, not TypeScript) are functionally equivalent:

```js
async function fn1() {
  await 1;
  return "x";
}

async function fn2() {
  await Promise.resolve(1);
  return "x";
}

async function fn3() {
  await 1;
  return Promise.resolve("x");
}

async function fn4() {
  await Promise.resolve(1);
  await Promise.resolve("x");
}
```

What we've chosen to do for TypeScript is implicitly unwrap the type for the promise. This is called the "awaited" type. The manner in which we do this is independent of the underlying promise implementation, allowing you to interact with promises from different libraries in a consistent manner. The code that you can write in TypeScript using async/await is no different than what you will be able to write in JavaScript (ES2015).

Given that we are following the specification with respect to parsing here, I'm going to close this issue. If you feel this is incorrect, you may choose to reopen this issue or file an issue against the [official proposal for ECMA-262](https://github.com/tc39/ecmascript-asyncawait).
